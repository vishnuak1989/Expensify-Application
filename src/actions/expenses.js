import uuid from 'uuid';
export const addExpense=({details="",note="",amount=0,createdAt=0} = {})=>({
  type:"ADD_EXPENSE",
  expense:{
    id:uuid(),
    details,
    note,
    amount,
    createdAt
  }

});
export const removeExpense=({id} = {})=>({
  type:"REMOVE_EXPENSE",
  expense:{
    id
  }

});

export const editExpense=(id,updates)=>({
  type:"EDIT_EXPENSE",
  id,
  updates

});
