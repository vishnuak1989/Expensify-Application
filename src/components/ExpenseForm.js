import React from 'react';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';


const now = moment();
console.log(now.format('MMM Do, YYYY'))
console.log(Date())
export default class ExpenseForm extends React.Component{
constructor(props){
  super(props);
  this.state={
    details:props.expense ? props.expense.details: "",
    note:props.expense ? props.expense.note: "",
    amount:props.expense ? (props.expense.amount/100).toString() : "",
    createdAt: props.expense ? moment(props.expense.createdAt): moment(),
    calenderFocused:false,
    error:""
        };
}
    onDetailsChange=(e)=>{
      const details = e.target.value;
      this.setState(()=>({details}));
    }
    onNoteChange=(e)=>{
      const note = e.target.value;
      this.setState(()=>({note}))
    }
    onAmountChange=(e)=>{
      const amount = e.target.value;
      if(!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)){
        this.setState(()=>({amount}))
      }
    }
   onDateChange=(createdAt)=>{
     if(createdAt){
      this.setState(()=>({createdAt}))
   }}
   onFocusChange=({focused})=>{
     this.setState(()=>({calenderFocused:focused}))
   }

   onSubmit=(e)=>{
     e.preventDefault();
     if (!this.state.details || !this.state.amount) {
      this.setState(()=>({error:"Please provide Details and Amount."}))
     }
     else{
       this.setState(()=>({error:""}))
       this.props.onSubmit({
         details:this.state.details,
         amount:parseFloat(this.state.amount,10) *100,
         createdAt:this.state.createdAt.valueOf(),
         note:this.state.note
       })
     }
   }

  render(){
    return (
      <div>
        <form onSubmit={this.onSubmit}>
          <input type="text" placeholder="Details" autoFocus value={this.state.details} onChange={this.onDetailsChange}/>
          <input type="text" placeholder="Amount" value={this.state.amount} onChange={this.onAmountChange}/>
          <SingleDatePicker
            date={this.state.createdAt}
            onDateChange={this.onDateChange}
            focused={this.state.calenderFocused}
            onFocusChange={this.onFocusChange}
            numberOfMonths={1}
            isOutsideRange={()=>false}
          />
          <textarea value={this.state.note} placeholder="Note about your expense(optional)" onChange={this.onNoteChange}>
          </textarea>
          <button>Save</button>
          {this.state.error && <p>{this.state.error}</p>}
        </form>
      </div>
    )
  }
}
