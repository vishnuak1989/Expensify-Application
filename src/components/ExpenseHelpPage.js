import React from 'react'

const ExpenseHelpPage=()=>(
  <div>
    <p>ExpenseHelpPage</p>
  </div>
)

export default ExpenseHelpPage;
