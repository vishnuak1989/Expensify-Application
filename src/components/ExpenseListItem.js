import React from 'react'
import {connect} from 'react-redux'
import{Link} from 'react-router-dom'

export const ExpenseListItem = ({id,details,amount,createdAt}) =>(

  <div>
    <Link to={`/edit/${id}`}><h3>{details}</h3></Link>
    <p>{amount} - {createdAt}</p>

  </div>
);

const mapStateToProps=(state)=>{
  return{
    expenses:state.expenses
  }
}


export default connect(mapStateToProps)(ExpenseListItem)
