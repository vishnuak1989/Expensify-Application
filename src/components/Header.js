import React from 'react'
import {NavLink } from 'react-router-dom'

const Header=()=>(
  <header>
    <h1>Expensify</h1>
    <NavLink to="/" activeClassName="selected" exact={true}>Dashboard</NavLink>
    <NavLink to="/create" activeClassName="selected">Create Expense</NavLink>
    {/* <NavLink to="/edit" activeClassName="selected">Edit Expense</NavLink> */}
    <NavLink to="/help" activeClassName="selected">Help</NavLink>
  </header>
)

export default Header;
