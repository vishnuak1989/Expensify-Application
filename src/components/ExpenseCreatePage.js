import React from 'react'
import ExpenseForm from './ExpenseForm'
import {connect} from 'react-redux'
import {addExpense} from '../actions/expenses'

 export class ExpenseCreatePage extends React.Component{
  onSubmit=(expense)=>{
    this.props.addExpense(expense)
    this.props.history.push('/')
  }
  render(){
  return(  <div>
      <h1>Create Expense</h1>
      <ExpenseForm onSubmit={this.onSubmit}/>
    </div>
  )
  }


}

const mapDispatchToProps=(dispatch)=>({
  addExpense:(expense) => dispatch(addExpense(expense))
})
export default connect(undefined,mapDispatchToProps)(ExpenseCreatePage);
