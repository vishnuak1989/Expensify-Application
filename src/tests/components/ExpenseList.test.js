import React from 'react'
import { ExpenseList } from '../../components/ExpenseList'
import Reactshallowrenderer from 'react-test-renderer/shallow';
import { shallow } from 'enzyme'
import expenses from '../fixtures/expenses'

test("Should render expense list with expenses",()=>{
  const wrapper = shallow(<ExpenseList expenses={expenses}/>);
  expect(wrapper).toMatchSnapshot();
})

test("Should render expense list with empty expenses",()=>{
  const wrapper = shallow(<ExpenseList expenses={[]} />);
  expect(wrapper).toMatchSnapshot();
})
