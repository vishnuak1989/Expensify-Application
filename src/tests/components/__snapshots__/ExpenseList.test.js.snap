// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render expense list with empty expenses 1`] = `
<div>
  <p>
     No expenses
  </p>
</div>
`;

exports[`Should render expense list with expenses 1`] = `
<div>
  <Connect(ExpenseListItem)
    amount={100}
    createdAt={0}
    details="Hello"
    id="1"
    key="1"
    note="This is a note"
  />
  <Connect(ExpenseListItem)
    amount={1000}
    createdAt={-345600000}
    details="Hello2"
    id="2"
    key="2"
    note="This is a note2"
  />
  <Connect(ExpenseListItem)
    amount={350}
    createdAt={345600000}
    details="Ola"
    id="3"
    key="3"
    note="This is a note3"
  />
</div>
`;
